Part 1: import requires libraries

In [4]:
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import numpy as np
from sklearn.linear_model import LinearRegression



data = sns.load_dataset('geyser')

Part 2: Creat function "machines_learning" to perform kfold method

In [10]:
def machine_learning(x, y, folds):
  #empty array for training and testing score
  training_score = []
  testing_score = []

  #use linear regression model
  model = LinearRegression(fit_intercept=True)

  #loops = how many folds
  for i in range(1,folds+1):
    #how many rows are index at once. caluclated by total row number/number of folds. 
    index = int(round(len(data)/folds,0))

    #if it's the last testing fold, we don't have to index it twice.
    if i == folds:
      testing = data[(i-1)*index : ]
      training = data[~data.isin(testing)].dropna()
   
    else:
      #testing equal data index from "fold number -1 * index number" to "fold number * index number". 
      #if i = 0 and folds = 10, data will be index from [0:27]
      testing = data[(i-1)*index :(i)*index]
      #training data = data -testingdata
      training = data[~data.isin(testing)].dropna()

    
    #Xtrain and Ytrain recieve training data with column name "x" or "y" to array. For x, reshape it to 2 columns.
    Xtrain = np.array(training[x]).reshape((len(training),1))
    Ytrain = np.array(training[y])
    #Xtrain and Ytrain recieve testing data with column name "x" or "y" to array. For x, reshape it to 2 columns.
    Xtest = np.array(testing[x]).reshape((len(testing),1))
    Ytest = np.array(testing[y])

    #fit model
    model.fit(Xtrain,Ytrain)
    #Append training and testing score for each loop into the stored array for training and testing
    training_score.append(model.score(Xtrain,Ytrain))
    testing_score.append(model.score(Xtest,Ytest))

  #print fold number
  print("When folds = " ,folds )
  #print result with mean&sd of training and testing score
  print('Training R^2 : mean = ',np.mean(training_score), "sd = " ,np.std(training_score))
  print('Testing R^2 : mean = ',np.mean(testing_score), "sd = " ,np.std(testing_score))





  
    
  

Part 3: Show off the function! 

In [12]:
machine_learning('waiting', 'duration', 10)
machine_learning('waiting', 'duration', 5)
machine_learning('waiting', 'duration', 3)
machine_learning('waiting', 'duration', 20)


When folds =  10
Training R^2 : mean =  0.8115509101080287 sd =  0.005225253310797004
Testing R^2 : mean =  0.8083525567740895 sd =  0.045430607147198473
When folds =  5
Training R^2 : mean =  0.8116959665550336 sd =  0.004389581073677457
Testing R^2 : mean =  0.8088129484571157 sd =  0.017030946649220467
When folds =  3
Training R^2 : mean =  0.8122519722140896 sd =  0.012034873772231723
Testing R^2 : mean =  0.8072394341897106 sd =  0.023847715261583875
When folds =  20
Training R^2 : mean =  0.8114890251337604 sd =  0.00365582610500259
Testing R^2 : mean =  0.801553996960115 sd =  0.0689459423833312


Part 4: Interpretation

  Based on our R^2(0.81155) scores, around 81% of the data fit the regression model we fit for  duration in minutes until the next eruption("waiting")  and the duration of an eruption in minutes (“duration”).In other words, the strength of relationship between "duration" and "waiting" is around 81%

  Repeated kfolds method also revealed that testing scores are almost identical(off by 1%) to the traiing score, which is fantastic. The relationship reveald by the training set is almost equivilent to the relationship of "waiting" and "duration" if we regenerate another set of data from the same DGP. And also, there are no overfitting.